subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DaemonBuildActionExecuter.java

```diff
@@ -40,7 +42,14 @@ public BuildActionResult execute(BuildAction action, ConnectionOperationParamete
         ClassPath classPath = DefaultClassPath.of(operationParameters.getInjectedPluginClasspath());
 
         DaemonParameters daemonParameters = parameters.getDaemonParameters();
-        BuildActionParameters actionParameters = new DefaultBuildActionParameters(daemonParameters.getEffectiveSystemProperties(), daemonParameters.getEnvironmentVariables(), SystemProperties.getInstance().getCurrentDir(), operationParameters.getBuildLogLevel(), daemonParameters.isEnabled(), classPath);
+        Map<String, String> effectiveSystemProperties;
+        Map<String, String> operationParametersSystemProperties = operationParameters.getSystemProperties(null);
```

Comment:
Instead of wiring this in here, would it make more sense to move this to `org.gradle.tooling.internal.provider.ProviderConnection#initParams` where all of the other parameter parsing happens?

Instead of having two `getEffectiveSystemProperties`, we would have one that returns mutable+immutable system properties always.

Here, we would always call `getEffectiveSystemProperties()`.

And then we just make sure that we set the mutable system properties in DaemonParameters with the system properties from `operationParameters.getSystemProperties(System.getProperties())`

Reply:
I've done what you asked for, but to me, the code seems even more twisted than before. PTAL.